File: 15.3sum.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
    ret = []
    nums.sort!
    for i in 0..nums.length-1
        if i > 0 and nums[i] == nums[i-1] then next end
        l, r = i+1, nums.length-1
        while l < r
            sum = nums[i] + nums[l] + nums[r]
            if sum < 0 then l += 1
            elsif sum > 0 then r -= 1
            else
                ret << [nums[i], nums[l], nums[r]]
                while l < r and nums[l] == nums[l + 1] do l += 1 end
                while l < r and nums[r] == nums[r - 1] do r -= 1 end
                l += 1
                r -= 1
            end
        end
    end
    return ret
end

File: 15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ret = []
        for j in range(1, n-1):
            if j > 1 and nums[j] == nums[j-1] and nums[j] == nums[j-2]:
                continue
            i, k = 0, n-1
            if nums[j] == nums[j-1]:
                i = j - 1
            while i < j and k > j:
                x = nums[i] + nums[j] + nums[k]
                if x < 0:
                    i += 1
                elif  x > 0:
                    k -= 1
                else:
                    ret.append([nums[i], nums[j], nums[k]])
                    i += 1
                    k -= 1
                    while i < j and nums[i] == nums[i-1]:
                        i += 1
                    while k > j and nums[k] == nums[k+1]:
                        k -= 1
        return ret